### PR TITLE
Improve Script parser errors and support for key parsing 

### PR DESCRIPTION
**Pull request by KtorZ**
- 25e199d5b8dc2e6b233556980b06761476032283
  :round_pushpin: **improve parsing error for 'Script' FromJSON instance**
    And in particular, run the validation also in the JSON parser. I've added a few negative test scenarios to show what errors look like.

- c4b9db9a8debd3882509dd82888efa8b1ea69e81
  :round_pushpin: **make 'keyHashFromText' works from key, extended keys and key hashes seamlessly.**
    This also impacts the command-line parser and the JSON parser to have the same behavior.

### CODE REVIEW

File: core/lib/Cardano/Address/Script.hs [hunk 1]
```diff
@@ -215,6 +228,72 @@ prettyErrKeyHashFromText = \case
...
         "Verification key hash is Bech32-encoded but has an invalid data part."
 
+--
+-- Script validation
+--
+
+-- | Validate a 'Script', semantically
+--
+-- @since 3.0.0
+validateScript :: Script -> Either ErrValidateScript ()
```
Comment (KtorZ) [inline core/lib/Cardano/Address/Script.hs:238]: moved from `Script.Parser`

File: core/lib/Cardano/Address/Script.hs [hunk 2]
```diff
@@ -215,6 +228,72 @@ prettyErrKeyHashFromText = \case
...
+  where
+    hasDuplicate xs = do
+        length sigs /= length (L.nub sigs)
+      where
+        sigs = [ sig | RequireSignatureOf sig <- xs ]
+
+-- | Possible validation errors when validating a script
+--
+-- @since 3.0.0
+data ErrValidateScript
```
Comment (KtorZ) [inline core/lib/Cardano/Address/Script.hs:267]: moved from `Script.Parser`

File: core/lib/Cardano/Address/Script.hs [hunk 3]
```diff
@@ -215,6 +228,72 @@ prettyErrKeyHashFromText = \case
...
+    | MZero
+    | DuplicateSignatures
+    | WrongKeyHash
+    | Malformed
+    deriving (Eq, Show)
+
+-- | Pretty-print a validation error.
+--
+-- @since 3.0.0
+prettyErrValidateScript
```
Comment (KtorZ) [inline core/lib/Cardano/Address/Script.hs:279]: moved from `Script.Parser`

File: command-line/test/Command/ScriptSpec.hs [hunk 4]
```diff
@@ -86,3 +95,12 @@ verKeyH2 = "script_vkh18srsxr3khll7vl3w9mqfu55n6wzxxlxj7qzr2mhnyrenxv223vj"
 
 verKeyH3 :: String
 verKeyH3 = "script_vkh18srsxr3khll7vl3w9mqfu55n6wzxxlxj7qzr2mhnyre5g2sfvk2"
+
+verExtKey4 :: String
+verExtKey4 = "script_xvk1mjr5lrrlxuvelx94hu2cttmg5pp6cwy5h0sa37qvpcd07pv9g23nlvugj5ez9qfxxvkmjwnpn69s48cv572phfy6qpmnwat0hwcdrasapqewe"
```
Comment (paweljakubas) [inline command-line/test/Command/ScriptSpec.hs:100]: :+1:

**Review by paweljakubas — APPROVED** [review]
LGTM

File: core/test/Cardano/Address/ScriptSpec.hs [hunk 5]
```diff
@@ -276,6 +276,52 @@ spec = do
...
+        it "List too small" $ do
+            let err = "Error in $: At least must not be larger than the list of keys."
+            Json.eitherDecode @Script
+                "{ \"some\":\
+                \   { \"at_least\": 2\
+                \   , \"from\":\
+                \       [\"script_vkh18srsxr3khll7vl3w9mqfu55n6wzxxlxj7qzr2mhnyreluzt36ms\"\
+                \       ]\
+                \   }\
+                \}" `shouldBe` Left err
```
Comment (piotr-iohk) [inline core/test/Cardano/Address/ScriptSpec.hs:304]: What is the valid input as a script here? readme says:
```
all [script_vkh15thv7uqqd0jh48fv4achd7ufqu9kv98c78r40q7fpwylw2c6wft,script_vkh18w06sur9ll7pcw8nzzr9mllgzyrrraaqse8336k2ej2cyph0u2a]
```
which is not a JSON?
As far as I see JSON is not supported?
```
{ "some": { "from": [ "script_vkh1yf07000d4ml3ywd3d439kmwp07xzgv6p35cwx8h605jfx0dtd4a", "script_vkh1mwlngj4fcwegw53tdmyemfupen2758xwvudmcz9ap8cnqk7jmh4", "script_vkh1qw4l62k4203dllrk3dk3sfjpnh3gufhtrtm4qvtrvn4xjp5x5rt" ], "at_least": 2 } }
```
:point_down: 

```
Parsing of the script failed. The script should be composed of nested lists, and the verification keys should be either encoded as bech32.
```

Comment (KtorZ) [reply core/test/Cardano/Address/ScriptSpec.hs:304]: >  As far as I see JSON is not supported?

Indeed, the command uses a custom DSL which is not JSON (but close). Arguably, we could support both. These particular tests are testing the JSON instance however. There are tests for the script DSL in https://github.com/input-output-hk/cardano-addresses/blob/master/core/test/Cardano/Address/Script/ParserSpec.hs#L40-L52

File: README.md [hunk 6]
```diff
@@ -25,198 +25,120 @@ API documentation is available [here](https://input-output-hk.github.io/cardano-
...
-  $ cat addr.prv \
-  | cardano-address key public --with-chain-code \
-  | cardano-address address payment --network-tag testnet \
-  | cardano-address address delegation $(cat stake.prv | cardano-address key public --with-chain-code)
-
-  addr_test1qq3a2mkasqfmsc5y5tguuckqaxa3pafrxs9x26d2hqvjehesytwfe2p0tt06y5ptxwvvm6lfu255dny7yz8e9kv4e83qwwzy29
+  <summary>How to generate a payment address from a payment key (<strong>payment.addr</strong>)</summary>
+  
+```console
+$ cardano-address address payment --network-tag testnet < payment.xvk > payment.addr
```
Comment (piotr-iohk) [inline README.md:85]: For consistency with previous examples (e.g. this one "How to generate a payment verification key (addr.xvk)"): 
```suggestion
$ cardano-address address payment --network-tag testnet < addr.xvk > payment.addr
```

**Review by piotr-iohk — APPROVED** [review]
:tada: 

Examples are very nice and clear. Now everyone, even myself, should be able to handle cardano-addresses with ease :)

**Pull request merged**